testing.md

```diff
@@ -0,0 +1,299 @@
+
+Example:
+
+Add this file to `test/unit_test.dart`:
+
+```dart
+import 'package:test/test.dart';
+
+void main() {
+  test('this is a unit test', () {
+    expect(42, 42);
+  });
+}
+```
+
+Run `flutter test test/unit_test.dart`.
+
+
+## Widget testing
+
+**NOTE: `flutter test` currently only works on Linux**
+
+A widget test is implemented just like a unit test. To let the test interact with
+the tested widget, Flutter provides a utility
+[`WidgetTester`](http://docs.flutter.io/flutter/flutter_test/WidgetTester-class.html).
+Use it to interact with the widget. For example, you can send tap and scroll
+gestures. Use it to find child widgets in the widget tree, read text, and verify
+that the values are correct.
+
+Example:
+
+Add this file to `test/widget_test.dart`:
+
+```dart
+import 'package:test/test.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+void main() {
+  test('this is a widget test', () {
+    // Creates a test environment for a widget
+    testWidgets((WidgetTester tester) {
+      // You can use keys to locate the widget you need to test
+      Key sliderKey = new UniqueKey();
+      double value = 0.0;
+
+      // Tells the tester to build a UI based on the widget tree passed to it
+      tester.pumpWidget(
+        new StatefulBuilder(
+          builder: (BuildContext context, StateSetter setState) {
+            return new Material(
+              child: new Center(
+                child: new Slider(
+                  key: sliderKey,
+                  value: value,
+                  onChanged: (double newValue) {
+                    setState(() {
+                      value = newValue;
+                    });
+                  }
+                )
+              )
+            );
+          }
+        )
+      );
+
+      expect(value, equals(0.0));
+
+      // Taps on the widget found by key
+      tester.tap(tester.findElementByKey(sliderKey));
+
+      // Verifies that the widget updated the value correctly
+      expect(value, equals(0.5));
+    });
+  });
+}
+```
+
+Run `flutter test test/widget_test.dart`.
+
+Check out [`package:flutter_test` API](http://docs.flutter.io/flutter/flutter_test/flutter_test-library.html)
+for all the utilities available for widget testing.
+
+
+## End-to-end testing
+
+A Flutter end-to-end test is also written using `package:test`. A full test is a
+pair - a test script and a Flutter application instrumented to receive commands
+from the test. Unlike unit and widget tests, end-to-end test code does not run
+in the same process as the application that's being tested. Instead, the tested
+application is launched on a _real device_ or in an _emulator_ (e.g. Android
+simulator or iOS Simulator). The test script runs on your computer. It connects
+to the application and issues commands to the application to perform various
+user actions. This is known as "driving" the application. Flutter provides tools
+and API, collectively referred to as _Flutter Driver_, to do just that.
```

Comment:
That depends on whether "I" stands for "interface" or "interfaces" :) Some people write "APIs", but I don't like that for some reason.

@qchong